internal/handlers/short.go

```diff
@@ -0,0 +1,87 @@
+package handlers
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"regexp"
+
+	"github.com/go-chi/chi/v5"
+)
+
+var jwtPayloadRegexp = regexp.MustCompile(`^.+\.(.+)\..+`)
+
+type payload struct {
+	Sub        string `json:"sub"`          // used by OCM tokens
+	InfraEnvID string `json:"infra_env_id"` // used by local auth tokens
+}
+
+// parseShortURL parses short-style URLs, where URL path segments are used to
+// hierarchically identify the desired resource. The URL ordering and structure
+// is defined on the router.
+func parseShortURL(r *http.Request) (*imageDownloadParams, int, error) {
+	imageID := chi.URLParam(r, "image_id")
+	token := chi.URLParam(r, "token")
+	if token == "" {
+		// try the "api_key" parameter name, which is used at the /byapikey/ path.
+		token = chi.URLParam(r, "api_key")
+	}
+	version := chi.URLParam(r, "version")
+	arch := chi.URLParam(r, "arch")
+	filename := chi.URLParam(r, "filename")
+
+	var err error
+	// the URL can have either the token or the image_id
+	if imageID == "" {
+		imageID, err = idFromJWT(token)
+		if err != nil {
+			return nil, http.StatusNotFound, err
+		}
+	}
+
+	params := imageDownloadParams{
+		imageID: imageID,
+		version: version,
+		arch:    arch,
+	}
+
+	switch filename {
+	case "minimal.iso":
+		params.imageType = "minimal-iso"
+	case "full.iso":
+		params.imageType = "full-iso"
+	default:
+		return nil, http.StatusNotFound, fmt.Errorf("unrecognized file name %s", filename)
+	}
+
+	return &params, 0, nil
+}
+
+func idFromJWT(jwt string) (string, error) {
```

Comment:
Maybe add a comment here that we acknowledge we're parsing and using the contents of the token without verifying it because that will be done by assisted later.

Also you can use the jwt library to do this for you (as we do in [assisted](https://github.com/openshift/assisted-service/blob/41251e4325c8faa041b139e2ee818de18969b5e5/internal/gencrypto/token.go#L93-L106)), but I'm just as happy to not add an additional dependency since the regex is rather simple and incredibly unlikely to change.

Reply:
I'll add a comment. And I agree that it's not worth pulling in a whole library to replace a few simple lines of code.